usr.bin/du/du.1
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)du.1	8.2 (Berkeley) 4/1/94
.\" $FreeBSD: src/usr.bin/du/du.1,v 1.15.2.9 2003/02/24 22:37:41 trhodes Exp $
.\"
.Dd April 1, 1994
.Dt DU 1
.Os
.Sh NAME
.Nm du
.Nd display disk usage statistics
.Sh SYNOPSIS
.Nm
.Op Fl H | L | P
.Op Fl I Ar mask
.Op Fl a | s | d Ar depth
.Op Fl c
.Op Fl h | k
.Op Fl x
.Op Fl t
.Op Ar
.Sh DESCRIPTION
The
.Nm
utility displays the file system block usage for each file argument
and for each directory in the file hierarchy rooted in each directory
argument.
If no file is specified, the block usage of the hierarchy rooted in
the current directory is displayed.
If the
.Fl k
flag is specified, the number of 1024-byte
blocks used by the file is displayed, otherwise
.Xr getbsize 3
is used to determine the preferred block size.
Partial numbers of blocks are rounded up.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl H
Symbolic links on the command line are followed, symbolic links in file
hierarchies are not followed.
.It Fl L
Symbolic links on the command line and in file hierarchies are followed.
.It Fl I Ar mask
Ignore files and directories matching the specified
.Ar mask .
.It Fl P
No symbolic links are followed.
This is the default.
.It Fl a
Display an entry for each file in a file hierarchy.
.It Fl h
"Human-readable" output.  Use unit suffixes: Byte, Kilobyte, Megabyte,
Gigabyte, Terabyte and Petabyte
.It Fl r
Generate messages about directories that cannot be read, files
that cannot be opened, and so on.  This is the default case.
This option exists solely for conformance with
.St -xpg4 .
.It Fl s
Display an entry for each specified file.
(Equivalent to
.Fl d Li 0 )
.It Fl t
Calculate the apparent size based on the file size (st_size) instead
of the actual media space used based on st_blkcnt.  This is meant to
give the caller an idea of how much data is present in actual transmittable
bytes if it was to be copied or sent elsewhere.
.Pp
To make this calculation
.Nm
uses st_size and assumes a media blocking granularity of 512 bytes.
Any holes present in the file are counted as transmittable data.
.It Fl d Ar depth
Display an entry for all files and directories
.Ar depth
directories deep.
.It Fl c
Display a grand total.
.It Fl k
Display block counts in 1024-byte (1-Kbyte) blocks.
.It Fl x
File system mount points are not traversed.
.El
.Pp
The
.Nm
utility counts the storage used by symbolic links and not the files they
reference unless the
.Fl H
or
.Fl L
option is specified.
If either the
.Fl H
or
.Fl L
options are specified, storage used by any symbolic links which are
followed is not counted or displayed.
.Pp
Files having multiple hard links are counted (and displayed) a single
time per
.Nm
execution.
.Sh ENVIRONMENT
.Bl -tag -width BLOCKSIZE
.It Ev BLOCKSIZE
If the environment variable
.Ev BLOCKSIZE
is set, and the
.Fl k
option is not specified, the block counts will be displayed in units of that
size block.
If
.Ev BLOCKSIZE
is not set, and the
.Fl k
option is not specified, the block counts will be displayed in 512-byte blocks.
.El
.Sh SEE ALSO
.Xr df 1 ,
.Xr fts 3 ,
.Xr symlink 7 ,
.Xr quot 8
.Sh HISTORY
A
.Nm
command appeared in
.At v1 .


usr.bin/du/du.c
/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Chris Newcomb.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * @(#) Copyright (c) 1989, 1993, 1994 The Regents of the University of California.  All rights reserved.
 * @(#)du.c	8.5 (Berkeley) 5/4/95
 * $FreeBSD: src/usr.bin/du/du.c,v 1.17.2.4 2002/12/12 16:29:39 trhodes Exp $
 */

#include <sys/param.h>
#include <sys/queue.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <fnmatch.h>
#include <fts.h>
#include <libutil.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>

#define	HASHSIZE	256		/* power of 2 only */
#define HASHMASK	(HASHSIZE - 1)

#define STBLOCKS(statp)	_stblocks(tflag, statp)

static SLIST_HEAD(ignhead, ignentry) ignores;
struct ignentry {
	char			*mask;
	SLIST_ENTRY(ignentry)	next;
};

static int	linkchk(FTSENT *);
static void	usage(void);
void		prthumanval(int64_t);
void		ignoreadd(const char *);
void		ignoreclean(void);
int		ignorep(FTSENT *);

static char period[] = ".";

typedef long long	du_number_t;

static blkcnt_t
_stblocks(int tflag, struct stat *st)
{
	if (tflag)
		return (st->st_size + 511) / 512;
	else
		return st->st_blocks;
}

int
main(int argc, char **argv)
{
	FTS		*fts;
	FTSENT		*p;
	long		blocksize;
	du_number_t	savednumber = 0;
	int		ftsoptions;
	int		listall;
	int		depth;
	int		Hflag, Lflag, Pflag;
	int		aflag, sflag, dflag, cflag, hflag, tflag;
	int		ch, notused, rval;
	char 		**save;

	Hflag = Lflag = Pflag = 0;
	aflag = sflag = dflag = cflag = hflag = tflag = 0;
	
	save = argv;
	ftsoptions = 0;
	depth = INT_MAX;
	SLIST_INIT(&ignores);
	
	while ((ch = getopt(argc, argv, "HI:LPastd:chkrx")) != -1)
		switch (ch) {
			case 'H':
				Hflag = 1;
				break;
			case 'I':
				ignoreadd(optarg);
				break;
			case 'L':
				if (Pflag)
					usage();
				Lflag = 1;
				break;
			case 'P':
				if (Lflag)
					usage();
				Pflag = 1;
				break;
			case 'a':
				aflag = 1;
				break;
			case 's':
				sflag = 1;
				break;
			case 't':
				tflag = 1;
				break;
			case 'd':
				dflag = 1;
				errno = 0;
				depth = atoi(optarg);
				if (errno == ERANGE || depth < 0) {
					warnx("invalid argument to option d: %s", optarg);
					usage();
				}
				break;
			case 'c':
				cflag = 1;
				break;
			case 'h':
				if (setenv("BLOCKSIZE", "512", 1) == -1)
					warn("setenv: cannot set BLOCKSIZE=512");
				hflag = 1;
				break;
			case 'k':
				hflag = 0;
				if (setenv("BLOCKSIZE", "1024", 1) == -1)
					warn("setenv: cannot set BLOCKSIZE=1024");
				break;
			case 'r':		 /* Compatibility. */
				break;
			case 'x':
				ftsoptions |= FTS_XDEV;
				break;
			case '?':
			default:
				usage();
		}

	argc -= optind;
	argv += optind;

	/*
	 * XXX
	 * Because of the way that fts(3) works, logical walks will not count
	 * the blocks actually used by symbolic links.  We rationalize this by
	 * noting that users computing logical sizes are likely to do logical
	 * copies, so not counting the links is correct.  The real reason is
	 * that we'd have to re-implement the kernel's symbolic link traversing
	 * algorithm to get this right.  If, for example, you have relative
	 * symbolic links referencing other relative symbolic links, it gets
	 * very nasty, very fast.  The bottom line is that it's documented in
	 * the man page, so it's a feature.
	 */

	if (Hflag + Lflag + Pflag > 1)
		usage();

	if (Hflag + Lflag + Pflag == 0)
		Pflag = 1;			/* -P (physical) is default */

	if (Hflag)
		ftsoptions |= FTS_COMFOLLOW;

	if (Lflag)
		ftsoptions |= FTS_LOGICAL;

	if (Pflag)
		ftsoptions |= FTS_PHYSICAL;

	listall = 0;

	if (aflag) {
		if (sflag || dflag)
			usage();
		listall = 1;
	} else if (sflag) {
		if (dflag)
			usage();
		depth = 0;
	}

	if (!*argv) {
		argv = save;
		argv[0] = period;
		argv[1] = NULL;
	}

	(void) getbsize(&notused, &blocksize);
	blocksize /= 512;

	rval = 0;
	
	if ((fts = fts_open(argv, ftsoptions, NULL)) == NULL)
		err(1, "fts_open");

	while ((p = fts_read(fts)) != NULL) {
		switch (p->fts_info) {
			case FTS_D:			/* Ignore. */
				if (ignorep(p))
					fts_set(fts, p, FTS_SKIP);
				break;
			case FTS_DP:
				if (ignorep(p))
					break;

				if (p->fts_pointer == NULL) {
					p->fts_pointer = malloc(sizeof(du_number_t));
					*(du_number_t *)p->fts_pointer = 0;
				}
				*(du_number_t *)p->fts_pointer += STBLOCKS(p->fts_statp);

				if (p->fts_parent->fts_pointer == NULL) {
					p->fts_parent->fts_pointer = malloc(sizeof(du_number_t));
					*(du_number_t *)p->fts_parent->fts_pointer = 0;
				}
				*(du_number_t *)p->fts_parent->fts_pointer += *(du_number_t *)p->fts_pointer += STBLOCKS(p->fts_statp);
				
				if (p->fts_level <= depth) {
					if (hflag) {
						(void) prthumanval(howmany(*(du_number_t *)p->fts_pointer, blocksize));
						(void) printf("\t%s\n", p->fts_path);
					} else {
					(void) printf("%lld\t%s\n",
					    howmany(*(du_number_t *)p->fts_pointer, blocksize),
					    p->fts_path);
					}
				}
				if (p->fts_pointer) {
					free(p->fts_pointer);
					p->fts_pointer = NULL;
				}
				break;
			case FTS_DC:			/* Ignore. */
				break;
			case FTS_DNR:			/* Warn, continue. */
			case FTS_ERR:
			case FTS_NS:
				warnx("%s: %s", p->fts_path, strerror(p->fts_errno));
				rval = 1;
				break;
			default:
				if (ignorep(p))
					break;

				if (p->fts_statp->st_nlink > 1 && linkchk(p))
					break;
				
				if (listall || p->fts_level == 0) {
					if (hflag) {
						(void) prthumanval(howmany(STBLOCKS(p->fts_statp),
							blocksize));
						(void) printf("\t%s\n", p->fts_path);
					} else {
						(void) printf("%lld\t%s\n",
							howmany((long long)STBLOCKS(p->fts_statp), blocksize),
							p->fts_path);
					}
				}
				if (p->fts_parent->fts_pointer == NULL) {
					p->fts_parent->fts_pointer = malloc(sizeof(du_number_t));
					*(du_number_t *)p->fts_parent->fts_pointer = 0;
				}
				*(du_number_t *)p->fts_parent->fts_pointer += STBLOCKS(p->fts_statp);
		}
		if (p->fts_parent->fts_pointer)
			savednumber = *(du_number_t *)p->fts_parent->fts_pointer;
	}

	if (errno)
		err(1, "fts_read");

	fts_close(fts);

	if (cflag) {
		if (hflag) {
			(void) prthumanval(howmany(savednumber, blocksize));
			(void) printf("\ttotal\n");
		} else {
			(void) printf("%lld\ttotal\n", howmany(savednumber, blocksize));
		}
	}

	ignoreclean();
	exit(rval);
}

static int
linkchk(FTSENT *p)
{
	struct links_entry {
		struct links_entry *next;
		struct links_entry *previous;
		int		links;
		dev_t		dev;
		ino_t		ino;
	};

	static const size_t links_hash_initial_size = 8192;
	static struct links_entry **buckets;
	static struct links_entry *free_list;
	static size_t number_buckets;
	static unsigned long number_entries;
	static char stop_allocating;
	struct links_entry *le, **new_buckets;
	struct stat *st;
	size_t i, new_size;
	int hash;

	st = p->fts_statp;

	/* If necessary, initialize the hash table. */
	if (buckets == NULL) {
		number_buckets = links_hash_initial_size;
		buckets = malloc(number_buckets * sizeof(buckets[0]));
		if (buckets == NULL)
			errx(1, "No memory for hardlink detection");
		for (i = 0; i < number_buckets; i++)
			buckets[i] = NULL;
	}

	/* If the hash table is getting too full, enlarge it. */
	if (number_entries > number_buckets * 10 && !stop_allocating) {
		new_size = number_buckets * 2;
		new_buckets = malloc(new_size * sizeof(struct links_entry *));

		/* Try releasing the free list to see if that helps. */
		if (new_buckets == NULL && free_list != NULL) {
			while (free_list != NULL) {
				le = free_list;
				free_list = le->next;
				free(le);
			}
			new_buckets = malloc(new_size * sizeof(new_buckets[0]));
		}

		if (new_buckets == NULL) {
			stop_allocating = 1;
			warnx("No more memory for tracking hard links");
		} else {
			memset(new_buckets, 0, new_size * sizeof(struct links_entry *));
			for (i = 0; i < number_buckets; i++) {
				while (buckets[i] != NULL) {
					/* Remove entry from old bucket. */
					le = buckets[i];
					buckets[i] = le->next;
	
					/* Add entry to new bucket. */
					hash = (le->dev ^ le->ino) % new_size;
	
					if (new_buckets[hash] != NULL)
						new_buckets[hash]->previous = le;
					le->next = new_buckets[hash];
					le->previous = NULL;
					new_buckets[hash] = le;
				}
			}
			free(buckets);
			buckets = new_buckets;
			number_buckets = new_size;
		}
	}

	/* Try to locate this entry in the hash table. */
	hash = ( st->st_dev ^ st->st_ino ) % number_buckets;
	for (le = buckets[hash]; le != NULL; le = le->next) {
		if (le->dev == st->st_dev && le->ino == st->st_ino) {
			/*
			 * Save memory by releasing an entry when we've seen
			 * all of it's links.
			 */
			if (--le->links <= 0) {
				if (le->previous != NULL)
					le->previous->next = le->next;
				if (le->next != NULL)
					le->next->previous = le->previous;
				if (buckets[hash] == le)
					buckets[hash] = le->next;
				number_entries--;
				/* Recycle this node through the free list */
				if (stop_allocating) {
					free(le);
				} else {
					le->next = free_list;
					free_list = le;
				}
			}
			return (1);
		}
	}

	if (stop_allocating)
		return (0);

	/* Add this entry to the links cache. */
	if (free_list != NULL) {
		/* Pull a node from the free list if we can. */
		le = free_list;
		free_list = le->next;
	} else
		/* Malloc one if we have to. */
		le = malloc(sizeof(struct links_entry));
	if (le == NULL) {
		stop_allocating = 1;
		warnx("No more memory for tracking hard links");
		return (0);
	}
	le->dev = st->st_dev;
	le->ino = st->st_ino;
	le->links = st->st_nlink - 1;
	number_entries++;
	le->next = buckets[hash];
	le->previous = NULL;
	if (buckets[hash] != NULL)
		buckets[hash]->previous = le;
	buckets[hash] = le;
	return (0);
}

void
prthumanval(int64_t bytes)
{
	char buf[sizeof("999M")];

	bytes *= DEV_BSIZE;

	humanize_number(buf, sizeof(buf), bytes, "", HN_AUTOSCALE,
			HN_B | HN_NOSPACE | HN_DECIMAL);

	(void) printf("%4s", buf);
}

static void
usage(void)
{
	(void)fprintf(stderr,
		"usage: du [-H | -L | -P] [-a | -s | -d depth] [-c] [-h | -k] [-x] [-I mask] [file ...]\n");
	exit(EX_USAGE);
}

void
ignoreadd(const char *mask)
{
	struct ignentry *ign;

	ign = calloc(1, sizeof(*ign));
	if (ign == NULL)
		errx(1, "cannot allocate memory");
	ign->mask = strdup(mask);
	if (ign->mask == NULL)
		errx(1, "cannot allocate memory");
	SLIST_INSERT_HEAD(&ignores, ign, next);
}

void
ignoreclean(void)
{
	struct ignentry *ign;
	
	while (!SLIST_EMPTY(&ignores)) {
		ign = SLIST_FIRST(&ignores);
		SLIST_REMOVE_HEAD(&ignores, next);
		free(ign->mask);
		free(ign);
	}
}

int
ignorep(FTSENT *ent)
{
	struct ignentry *ign;

	SLIST_FOREACH(ign, &ignores, next)
		if (fnmatch(ign->mask, ent->fts_name, 0) != FNM_NOMATCH)
			return 1;
	return 0;
}
